gpflow/posteriors.py

```diff
@@ -381,66 +381,43 @@ def _conditional_with_precompute(
         Relies on cached alpha and Qinv.
         """
         assert_params_false(self._conditional_with_precompute, full_output_cov=full_output_cov)
+        err, Lm = cache
 
-        (alpha,) = cache
-        (Qinv,) = cache
-
-        Kmn = self.kernel(self.X_data, Xnew)
-        # compute kernel stuff
-        num_func = tf.shape(self.Y_data)[-1]  # R
-        N = tf.shape(Kmn)[-1]
-
-        # get the leading dims in Kmn to the front of the tensor Kmn
-        K = tf.rank(Kmn)
-        perm = tf.concat(
-            [
-                tf.reshape(tf.range(1, K - 1), [K - 2]),  # leading dims (...)
-                tf.reshape(0, [1]),  # [M]
-                tf.reshape(K - 1, [1]),
-            ],
-            0,
-        )  # [N]
-        Kmn = tf.transpose(Kmn, perm)  # [..., M, N]
-        leading_dims = tf.shape(Kmn)[:-2]
-
-        # get the leading dims in Knm to the front of the tensor Knm
-        Knm = leading_transpose(Kmn, [..., -1, -2])
-
-        assert self.mean_function is not None
         Knn = self.kernel(Xnew, full_cov=full_cov)
-        err = self.Y_data - self.mean_function(self.X_data)
-
-        mean = Knm @ alpha @ err
-
-        # The GPR model only has a single latent GP.
-        if full_cov:
-            cov = Knn - Knm @ Qinv @ Kmn  # [..., N, N]
-            cov_shape = tf.concat([leading_dims, [num_func, N, N]], 0)
-            cov = tf.broadcast_to(tf.expand_dims(cov, -3), cov_shape)  # [..., R, N, N]
-
-        else:
-            cov = Knn - tf.einsum("...ij,...ji->...i", Knm @ Qinv, Kmn)  # [..., N]
-            cov_shape = tf.concat([leading_dims, [num_func, N]], 0)  # [..., R, N]
-            cov = tf.broadcast_to(tf.expand_dims(cov, -2), cov_shape)  # [..., R, N]
-            cov = tf.linalg.adjoint(cov)
+        Kmn = self.kernel(self.X_data, Xnew)
 
-        return mean, cov
+        return base_conditional_with_lm(
+            Kmn=Kmn,
+            Lm=Lm,
+            Knn=Knn,
+            f=err,
+            full_cov=full_cov,
+            q_sqrt=None,
+            white=False,
+        )
 
     @check_shapes(
-        "return[0].value: [M, M]",
+        "return[0].value: [M, D]",
```

Comment:
Suggestion: you could register the shape of a `PrecomputedValue` for `check_shapes`...

Reply:
> Suggestion: you could register the shape of a `PrecomputedValue` for `check_shapes`...

Done.